src/inmanta/deploy/tasks.py

```diff
@@ -0,0 +1,182 @@
+class Task(abc.ABC):
+    """
+    Resource action task. Represents the execution of a specific resource action for a given resource.
+    """
+
+    resource: ResourceIdStr
+
+    @abc.abstractmethod
+    async def execute(self, scheduler: "scheduler.ResourceScheduler", agent: str) -> None:
+        """the scheduler is considered to be a friend class: access to internal members is expected"""
+        pass
+
+    def delete_with_resource(self) -> bool:
+        return True
+
+
+class PoisonPill(Task):
+    """
+    Task to signal queue shutdown
+
+    It is used to make sure all workers wake up to observe that they have been closed.
+    It functions mostly as a no-op
+    """
+
+    async def execute(self, scheduler: "scheduler.ResourceScheduler", agent: str) -> None:
+        pass
+
+
+class WithHashMatchTask(Task):
+
+    async def execute(self, scheduler: "scheduler.ResourceScheduler", agent: str) -> None:
+        resource_details: "state.ResourceDetails"
+        async with scheduler._scheduler_lock:
+            # fetch resource details atomically under lock
+            try:
+                resource_details = scheduler._state.resources[self.resource]
+            except KeyError:
+                # Stale resource, can simply be dropped.
+                # May occur in rare races between new_version and acquiring the lock we're under here. This race is safe
+                # because of this check, and an intrinsic part of the locking design because it's preferred over wider
+                # locking for performance reasons.
+                return
+        await self.execute_on_resource(scheduler, agent, resource_details)
+
+    @abc.abstractmethod
+    async def execute_on_resource(
+        self, scheduler: "scheduler.ResourceScheduler", agent: str, resource_details: "state.ResourceDetails"
+    ) -> None:
+        pass
+
+
+class Deploy(WithHashMatchTask):
+    async def execute_on_resource(
+        self, scheduler: "scheduler.ResourceScheduler", agent: str, resource_details: "state.ResourceDetails"
+    ) -> None:
+        status = await self.do_deploy(scheduler, agent, resource_details)
+
+        is_success = status == const.ResourceState.deployed
+
+        async with scheduler._scheduler_lock:
+            # refresh resource details for latest model state
+            new_details: Optional[state.ResourceDetails] = scheduler._state.resources.get(self.resource, None)
+            my_state: state.ResourceState | None = scheduler._state.resource_state.get(self.resource, None)
+            if new_details is not None and new_details.attribute_hash == resource_details.attribute_hash:
+                assert my_state is not None
+                if is_success:
+                    my_state.status = state.ResourceStatus.UP_TO_DATE
+                    my_state.deployment_result = state.DeploymentResult.DEPLOYED
+                else:
+                    # FIXME: WDB to Sander: do we set status here as well?
```

Comment:
According to the current definition of the status yes, but there's a FIXME attached to that as well, so perhaps we can mark this accordingly?
```suggestion
                    # FIXME[#8008]: WDB to Sander: do we set status here as well?
```